pkg/controller/direct/compute/forwardingrule_controller.go

```diff
@@ -0,0 +1,460 @@
+		// Get the created resource
+		created, err = a.forwardingRulesClient.Get(ctx, getReq)
+		if err != nil {
+			if direct.IsNotFound(err) {
+				return err
+			}
+			return fmt.Errorf("getting ComputeForwardingRule %q failed: %w", a.resourceID, err)
+		}
+	}
+
+	status := &krm.ComputeForwardingRuleStatus{
+		CreationTimestamp: op.Proto().CreationTimestamp,
+		SelfLink:          created.SelfLink,
+	}
+	return setStatus(u, status)
+}
+
+func (a *forwardingRuleAdapter) Update(ctx context.Context, u *unstructured.Unstructured) error {
+	if a.resourceID == "" {
+		return fmt.Errorf("resourceID is empty")
+	}
+
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("updating ComputeForwardingRule", "name", a.resourceID)
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+	forwardingRule := ComputeForwardingRuleSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	forwardingRule.Name = direct.LazyPtr(a.resourceID)
+	forwardingRule.Labels = desired.Labels
+
+	// Patch only support update on networkTier field, which KCC does not support yet.
+	// Use setTarget and setLabels to update target and labels fields.
+	var err error
+	// TODO(yuhou): Checked the realGCP logs, setLabels request is being sent even when there are no updates to labels.
+	// That might because of the generated labelsFingerPrint?
+	if a.Location == "global" {
+		setLabelsReq := &computepb.SetLabelsGlobalForwardingRuleRequest{
+			Resource:                       a.resourceID,
+			GlobalSetLabelsRequestResource: &computepb.GlobalSetLabelsRequest{LabelFingerprint: a.actual.LabelFingerprint, Labels: forwardingRule.Labels},
+			Project:                        a.projectID,
+		}
+		_, err = a.globalForwardingRulesClient.SetLabels(ctx, setLabelsReq)
+	} else {
+		setLabelsReq := &computepb.SetLabelsForwardingRuleRequest{
+			Resource:                       a.resourceID,
+			RegionSetLabelsRequestResource: &computepb.RegionSetLabelsRequest{LabelFingerprint: a.actual.LabelFingerprint, Labels: forwardingRule.Labels},
+			Project:                        a.projectID,
+			Region:                         a.Location,
+		}
+		_, err = a.forwardingRulesClient.SetLabels(ctx, setLabelsReq)
+	}
+	if err != nil {
+		return fmt.Errorf("updating ComputeForwardingRule labels %s failed: %w", a.resourceID, err)
+	}
+	log.V(2).Info("successfully updated ComputeForwardingRule labels", "name", a.resourceID)
+
+	// setTarget request is sent when there are updates to target.
+	if !reflect.DeepEqual(forwardingRule.Target, a.actual.Target) {
+		if a.Location == "global" {
+			setTargetReq := &computepb.SetTargetGlobalForwardingRuleRequest{
+				ForwardingRule:          a.resourceID,
+				TargetReferenceResource: &computepb.TargetReference{Target: forwardingRule.Target},
+				Project:                 a.projectID,
+			}
+			_, err = a.globalForwardingRulesClient.SetTarget(ctx, setTargetReq)
```

Comment:
As before, we need some way to be able to look at the code and know that this is not returning an operation.

Reply:
It returns an operation. I added a check for op.Wait()